File: src/tetris.c
/******************************************************************************
 * Copyright (C) 2016  Renan S. Silva                                         *
 *                                                                            *
 * This software is provided 'as-is', without any express or implied          *
 * warranty. In no event will the authors be held liable for any damages      *
 * arising from the use of this software.                                     *
 *                                                                            *
 * Permission is granted to anyone to use this software for any purpose,      *
 * including commercial applications, and to alter it and redistribute it     *
 * freely, subject to the following restrictions:                             *
 *                                                                            *
 * 1. The origin of this software must not be misrepresented; you must not    *
 *    claim that you wrote the original software. If you use this software    *
 *    in a product, an acknowledgment in the product documentation would be   *
 *    appreciated but is not required.                                        *
 * 2. Altered source versions must be plainly marked as such, and must not be *
 *    misrepresented as being the original software.                          *
 * 3. This notice may not be removed or altered from any source distribution. *
 ******************************************************************************/
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "other_window.h"
#include "types.h"
#include "tetris.h"
#include "lelmark.h"
#include "trainer.h"
#include "graphics.h"

_piece Ta_piece     = {{ -1, -1 }, {  0, -1 }, { 1, -1 }, { 0, 0 }} ; // Ta
_piece Tb_piece     = {{  0, -2 }, {  0, -1 }, { 1, -1 }, { 0, 0 }} ; // Tb
_piece Tc_piece     = {{ -1, -1 }, { -2,  0 }, {-1,  0 }, { 0, 0 }} ; // Tc
_piece Td_piece     = {{  0, -2 }, { -1, -1 }, { 0, -1 }, { 0, 0 }} ; // Td

_piece Ja_piece     = {{ -2, -1 }, { -1, -1 }, { 0, -1 }, { 0, 0 }} ; // Ja
_piece Jb_piece     = {{  0, -2 }, {  1, -2 }, { 0, -1 }, { 0, 0 }} ; // Jb
_piece Jc_piece     = {{ -2, -1 }, { -2,  0 }, {-1,  0 }, { 0, 0 }} ; // Jc
_piece Jd_piece     = {{  0, -2 }, {  0, -1 }, {-1,  0 }, { 0, 0 }} ; // Jd

_piece La_piece     = {{  0, -1 }, {  1, -1 }, { 2, -1 }, { 0, 0 }} ; // La
_piece Lb_piece     = {{ -1, -2 }, { -1, -1 }, {-1,  0 }, { 0, 0 }} ; // Lb
_piece Lc_piece     = {{  0, -1 }, { -2,  0 }, {-1,  0 }, { 0, 0 }} ; // Lc
_piece Ld_piece     = {{ -1, -2 }, {  0, -2 }, { 0, -1 }, { 0, 0 }} ; // Ld

_piece Sa_piece     = {{  0, -1 }, {  1, -1 }, {-1,  0 }, { 0, 0 }} ; // Sa
_piece Sb_piece     = {{ -1, -2 }, { -1, -1 }, { 0, -1 }, { 0, 0 }} ; // Sb

_piece Ia_piece     = {{ -3, 0  }, { -2,  0 }, {-1,  0 }, { 0, 0 }} ; // Ia
_piece Ib_piece     = {{  0, -3 }, {  0, -2 }, { 0, -1 }, { 0, 0 }} ; // Ib

_piece Za_piece     = {{ -2, -1 }, { -1, -1 }, {-1,  0 }, { 0, 0 }} ; // Za
_piece Zb_piece     = {{  1, -2 }, {  0, -1 }, { 1, -1 }, { 0, 0 }} ; // Zb

_piece Square_piece = {{ -1, -1 }, {  0, -1 }, {-1,  0 }, { 0, 0 }} ; // Square

int cleaned_any_row() {
    _bg_info *bg_info = get_bg_info_pointer();

    for (int j = 0; j < __Y_SIZE; ++j) {
        int ok = 1;
        for (int i = 0; i < __X_SIZE; ++i) {
            if ( bg_info->data[i][j] == 0 ) {
                ok = 0;
                break;
            }
        }

        if ( ok ) {
            return 1;
        }
    }

    return 0;
}

int piece_touched_the_ground (_piece piece, int dx, int dy){
    int x = get_cpu_pointer()->mem_controller.memory[0xff92] - 8;
    int y = get_cpu_pointer()->mem_controller.memory[0xff93] - 16;

    int max_y = -999;

    dx += x;

    x = dx;

    max_y = max_y > piece.a.y ? max_y : piece.a.y;
    max_y = max_y > piece.b.y ? max_y : piece.b.y;
    max_y = max_y > piece.c.y ? max_y : piece.c.y;
    max_y = max_y > piece.d.y ? max_y : piece.d.y;

    max_y *= 8;

    if ( y + dy + max_y >= 142 )
        return 1;

    return 0;
}

int is_inside_bounds(_piece piece, int dx, int dy){
    int x = get_cpu_pointer()->mem_controller.memory[0xff92] - 8;
    int y = get_cpu_pointer()->mem_controller.memory[0xff93] - 16;

    int min_x =  999;
    int max_x = -999;

    dy += y;

    y = dy;

    min_x = min_x < piece.a.x ? min_x : piece.a.x;
    min_x = min_x < piece.b.x ? min_x : piece.b.x;
    min_x = min_x < piece.c.x ? min_x : piece.c.x;
    min_x = min_x < piece.d.x ? min_x : piece.d.x;

    max_x = max_x > piece.a.x ? max_x : piece.a.x;
    max_x = max_x > piece.b.x ? max_x : piece.b.x;
    max_x = max_x > piece.c.x ? max_x : piece.c.x;
    max_x = max_x > piece.d.x ? max_x : piece.d.x;

    min_x *= 8;
    max_x *= 8;

    if ( x + dx + min_x < 16 || x + dx + max_x > 88 ) {
        return 0;
    }

    return 1;
}

int add_piece(_piece piece, int dx, int dy) {
    int x = get_cpu_pointer()->mem_controller.memory[0xff92] - 8;
    int y = get_cpu_pointer()->mem_controller.memory[0xff93] - 16;
    _bg_info *bg = get_bg_info_pointer();

    int xx;
    int yy;

    xx = (x + dx + piece.a.x*8) / 8 - 2;
    yy = (y + dy + piece.a.y*8) / 8 - 1;
    bg->data[xx][yy] += 2;

    xx = (x + dx + piece.b.x*8) / 8 - 2;
    yy = (y + dy + piece.b.y*8) / 8 - 1;
    bg->data[xx][yy] += 2;

    xx = (x + dx + piece.c.x*8) / 8 - 2;
    yy = (y + dy + piece.c.y*8) / 8 - 1;
    bg->data[xx][yy] += 2;

    xx = (x + dx + piece.d.x*8) / 8 - 2;
    yy = (y + dy + piece.d.y*8) / 8 - 1;
    bg->data[xx][yy] += 2;

    return 1;
}

int can_fit(_piece piece, int dx, int dy) {
    int x = get_cpu_pointer()->mem_controller.memory[0xff92] - 8;
    int y = get_cpu_pointer()->mem_controller.memory[0xff93] - 16;
    _bg_info *bg = get_bg_info_pointer();

    int xx ;
    int yy ;

    xx = (x + dx + piece.a.x*8) / 8 - 2;
    yy = (y + dy + piece.a.y*8) / 8 - 1;

    if ( bg->data[xx][yy] == 1 ) return 0;

    xx = (x + dx + piece.b.x*8) / 8 - 2;
    yy = (y + dy + piece.b.y*8) / 8 - 1;

    if ( bg->data[xx][yy] == 1 ) return 0;

    xx = (x + dx + piece.c.x*8) / 8 - 2;
    yy = (y + dy + piece.c.y*8) / 8 - 1;

    if ( bg->data[xx][yy] == 1 ) return 0;

    xx = (x + dx + piece.d.x*8) / 8 - 2;
    yy = (y + dy + piece.d.y*8) / 8 - 1;

    if ( bg->data[xx][yy] == 1 ) return 0;

    return 1;
}

void clear_lines() {
    _bg_info *bg_info = get_bg_info_pointer();

    for (int j = 0; j < __Y_SIZE; ++j) {
        int ok = 1;
        for (int i = 0; i < __X_SIZE; ++i) {
            if ( bg_info->data[i][j] == 0 ) {
                ok = 0;
                break;
            }
        }

        if ( ok ) {
            for (int i = 0; i < __X_SIZE; ++i) {
                bg_info->data[i][j] = 5;
            }
        }
    }

    int changed;

    /*printf("\n*****************************\n\nCleared line\n");*/

    do {
        /*dump_bg();*/
        changed = 0;
        for (int j = 0; j < __Y_SIZE; ++j) {
            if ( bg_info->data[0][j] == 5 ) {
                for (int i = j; i > 0; i--) {
                    for (int k = 0; k < __X_SIZE; k++) {
                        bg_info->data[k][i] = bg_info->data[k][i-1];
                        changed = 1;
                    }
                }
            }
        }
    } while ( changed );
    /*dump_bg();*/
}

void save_bg() {
    _bg_info *bg_bkp = &(get_brain_pointer()->bg_info_copy);
    _bg_info *bg = get_bg_info_pointer();

    memcpy(bg_bkp, bg, sizeof(_bg_info));
}

void restore_bg() {
    _bg_info *bg_bkp = &(get_brain_pointer()->bg_info_copy);
    _bg_info *bg = get_bg_info_pointer();

    memcpy(bg, bg_bkp, sizeof(_bg_info));
}

void dump_bg() {
    printf("    +---------------------+\n");
    for (int j = 0; j < __Y_SIZE; ++j) {
        printf("%3d | ", j);
        for (int i = 0; i < __X_SIZE; ++i) {
            switch ( get_bg_info_pointer()->data[i][j] ) {
                case 0:
                    printf("  ");
                    break;
                case 1:
                    printf("X ");
                    break;
                case 2:
                    printf("O ");
                    break;
                case 5:
                    printf("%% ");
                    break;
                default:
                    printf("? ");
                    break;
            }
        }
        printf("|\n");
    }
    printf("    +---------------------+\n\n");
}

void reset_bg() {
    _bg_info *bg_info = get_bg_info_pointer();

    for (int i = 0; i < __X_SIZE; i++) {
        for (int j = 0; j < __Y_SIZE; j++) {
             bg_info->data[i][j] = 0;
        }
    }
}

void get_best_move(){
    double best_cost = -999999;
    _best_piece *best_piece = get_best_piece_pointer();

    best_piece->set = 0;

    _piece_type piece_type = get_current_piece();
    _piece piece = get_piece_coord_from_id(piece_type);

    int x = get_cpu_pointer()->mem_controller.memory[0xff92] - 8;
    int y = get_cpu_pointer()->mem_controller.memory[0xff93] - 24;

    for (int n_totation = 0; n_totation < get_piece_rotation(piece_type); ++n_totation) {
        for (int dx = -80 ; dx < 96; dx += 8) {
            if ( is_inside_bounds(piece, dx, 16)) {
                int first = 0;
                for (int dy = 24; dy < 8*20; dy += 8 ) {
                    if ( can_fit(piece, dx, dy )) {
                        first = 1;
                    } else if ( first ) {
                        save_bg();
                        add_piece(piece, dx, dy - 8);
                        evaluate_cost();
                        if ( best_cost < get_cost() ) {
                            best_cost           = get_cost();
                            best_piece->coord.x = dx + x;
                            best_piece->coord.y = dy + y;
                            best_piece->type    = piece_type;
                            best_piece->blocks  = piece;
                            best_piece->set     = 1;
                            best_piece->nrotations = n_totation;
                            best_piece->parameters = get_brain_pointer()->population[get_brain_pointer()->current];
                        }

                        restore_bg();
                        break;
                    } else {
                        break;
                    }

                    if ( piece_touched_the_ground (piece, dx, dy) ) {
                        save_bg();
                        add_piece(piece, dx, dy - 8);
                        evaluate_cost();
                        best_cost           = get_cost();
                        best_piece->coord.x = dx + x;
                        best_piece->coord.y = dy + y;
                        best_piece->type    = piece_type;
                        best_piece->blocks  = piece;
                        best_piece->set     = 1;
                        best_piece->nrotations = n_totation;
                        best_piece->parameters = get_brain_pointer()->population[get_brain_pointer()->current];
                        restore_bg();
                        break;
                    }
                }
            }
        }

        piece_type = rotate_piece ( piece_type );
        piece = get_rotated_piece ( piece_type );
    }
}

_piece get_rotated_piece (_piece_type piece_type ) {
    switch (piece_type) {
        case La:
            return La_piece;
        case Lb:
            return Lb_piece;
        case Lc:
            return Lc_piece;
        case Ld:
            return Ld_piece;

        case Ta:
            return Ta_piece;
        case Tb:
            return Tb_piece;
        case Tc:
            return Tc_piece;
        case Td:
            return Td_piece;

        case Ja:
            return Ja_piece;
        case Jb:
            return Jb_piece;
        case Jc:
            return Jc_piece;
        case Jd:
            return Jd_piece;

        case Za:
            return Za_piece;
        case Zb:
            return Zb_piece;

        case Sa:
            return Sa_piece;
        case Sb:
            return Sb_piece;

        case Ia:
            return Ia_piece;
        case Ib:
            return Ib_piece;

        case SQUARE:
            return Square_piece;

        default:
            fprintf(stderr, "Invalid piece in rotate_piece\n");
            abort();
    }
}

_piece_type rotate_piece (_piece_type piece_type ) {
    switch (piece_type) {
        case La:
            return Lb;
        case Lb:
            return Lc;
        case Lc:
            return Ld;
        case Ld:
            return La;

        case Ta:
            return Tb;
        case Tb:
            return Tc;
        case Tc:
            return Td;
        case Td:
            return Ta;

        case Ja:
            return Jb;
        case Jb:
            return Jc;
        case Jc:
            return Jd;
        case Jd:
            return Ja;

        case Za:
            return Zb;
        case Zb:
            return Za;

        case Sa:
            return Sb;
        case Sb:
            return Sa;

        case Ia:
            return Ib;
        case Ib:
            return Ia;

        case SQUARE:
            return SQUARE;

        default:
            fprintf(stderr, "Invalid piece in rotate_piece\n");
            abort();
    }
}

int get_piece_rotation ( _piece_type piece_type ) {
    switch (piece_type) {
        case La:
        case Lb:
        case Lc:
        case Ld:

        case Ta:
        case Tb:
        case Tc:
        case Td:

        case Ja:
        case Jb:
        case Jc:
        case Jd:
            return 4;

        case Za:
        case Zb:

        case Sa:
        case Sb:

        case Ia:
        case Ib:
            return 2;

        case SQUARE:
            return 1;

        default:
            fprintf(stderr, "Invalid piece on get_piece_rotation\n");
            abort();
    }
}

_piece get_piece_coord_from_id( _piece_type piece_type ) {
    switch (piece_type) {
        case La:
        case Lb:
        case Lc:
        case Ld:
            return La_piece;

        case Ta:
        case Tb:
        case Tc:
        case Td:
            return Ta_piece;

        case Ja:
        case Jb:
        case Jc:
        case Jd:
            return Ja_piece;

        case Za:
        case Zb:
            return Za_piece;

        case Sa:
        case Sb:
            return Sa_piece;

        case Ia:
        case Ib:
            return Ia_piece;

        case SQUARE:
            return Square_piece;

        default:
            fprintf(stderr, "Invalid piece in get_piece_coord_from_id\n");
            abort();
    }
}

_piece_type get_current_piece(){
    _cpu_info *cpu = get_cpu_pointer();
    int index;
    int value;

    index = 0xc203;
    value = cpu->mem_controller.memory[index];

    switch (value) {
        case 0x0c:
        case 0x0d:
        case 0x0e:
        case 0x0f:
            return SQUARE;

        case 0x04:
            return Ja;
        case 0x05:
            return Jb;
        case 0x06:
            return Jc;
        case 0x07:
            return Jd;

        case 0x00:
            return La;
        case 0x01:
            return Lb;
        case 0x02:
            return Lc;
        case 0x03:
            return Ld;

        case 0x08:
        case 0x0a:
            return Ia;
        case 0x09:
        case 0x0b:
            return Ib;

        case 0x14:
        case 0x16:
            return Sa;
        case 0x15:
        case 0x17:
            return Sb;

        case 0x10:
        case 0x12:
            return Za;
        case 0x11:
        case 0x13:
            return Zb;

        case 0x18:
            return Ta;
        case 0x19:
            return Tb;
        case 0x1a:
            return Tc;
        case 0x1b:
            return Td;

        default:
            fprintf(stderr, "Invalid piece in get_current_piece\n");
            abort();
    }
}

File: src/tetris.jl
include("j_utils.jl")
include("input_handler.jl")
include("types.jl")
include("feature_functions.jl")
include("position_evaluator.jl")
include("draw.jl")

brain = ga_brain()
gs    = game_state() #init_game_state()

Ta_piece = [ -1 -1 ;  0 -1 ;  1 -1 ; 0 0 ] # Ta
Tb_piece = [  0 -2 ;  0 -1 ;  1 -1 ; 0 0 ] # Tb
Tc_piece = [ -1 -1 ; -2  0 ; -1  0 ; 0 0 ] # Tc
Td_piece = [  0 -2 ; -1 -1 ;  0 -1 ; 0 0 ] # Td

Ja_piece = [ -2 -1 ; -1 -1 ;  0 -1 ; 0 0 ] # Ja
Jb_piece = [  0 -2 ;  1 -2 ;  0 -1 ; 0 0 ] # Jb
Jc_piece = [ -2 -1 ; -2  0 ; -1  0 ; 0 0 ] # Jc
Jd_piece = [  0 -2 ;  0 -1 ; -1  0 ; 0 0 ] # Jd

La_piece = [  0 -1 ;  1 -1 ;  2 -1 ; 0 0 ] # La
Lb_piece = [ -1 -2 ; -1 -1 ; -1  0 ; 0 0 ] # Lb
Lc_piece = [  0 -1 ; -2  0 ; -1  0 ; 0 0 ] # Lc
Ld_piece = [ -1 -2 ;  0 -2 ;  0 -1 ; 0 0 ] # Ld

Sa_piece = [  0 -1 ;  1 -1 ; -1  0 ; 0 0 ] # Sa
Sb_piece = [ -1 -2 ; -1 -1 ;  0 -1 ; 0 0 ] # Sb

Ia_piece = [ -3  0 ; -2  0 ; -1  0 ; 0 0 ] # Ia
Ib_piece = [  0 -3 ;  0 -2 ;  0 -1 ; 0 0 ] # Ib

Za_piece = [ -2 -1 ; -1 -1 ; -1  0 ; 0 0 ] # Za
Zb_piece = [  1 -2 ;  0 -1 ;  1 -1 ; 0 0 ] # Zb

O_piece  = [ -1 -1 ;  0 -1 ; -1  0 ; 0 0 ] # O

function get_piece_n_rotations(piece :: Int64)
    d = Dict{Int64, Int64}()

    d[0x00] = 4
    d[0x01] = 4
    d[0x02] = 4
    d[0x03] = 4

    d[0x04] = 4
    d[0x05] = 4
    d[0x06] = 4
    d[0x07] = 4

    d[0x08] = 2
    d[0x09] = 2
    d[0x0a] = 2
    d[0x0b] = 2

    d[0x0c] = 1
    d[0x0d] = 1
    d[0x0e] = 1
    d[0x0f] = 1

    d[0x10] = 2
    d[0x11] = 2
    d[0x12] = 2
    d[0x13] = 2

    d[0x14] = 2
    d[0x15] = 2
    d[0x16] = 2
    d[0x17] = 2

    d[0x18] = 4
    d[0x19] = 4
    d[0x1a] = 4
    d[0x1b] = 4

    return d[piece]
end

function get_piece_offsets(piece :: Int64)
    d = Dict{Int64, Array{Int64, 2}}()

    d[0x00] = La_piece
    d[0x01] = Lb_piece
    d[0x02] = Lc_piece
    d[0x03] = Ld_piece

    d[0x04] = Ja_piece
    d[0x05] = Jb_piece
    d[0x06] = Jc_piece
    d[0x07] = Jd_piece

    d[0x08] = Ia_piece
    d[0x09] = Ib_piece
    d[0x0a] = Ia_piece
    d[0x0b] = Ib_piece

    d[0x0c] = O_piece
    d[0x0d] = O_piece
    d[0x0e] = O_piece
    d[0x0f] = O_piece

    d[0x10] = Za_piece
    d[0x11] = Zb_piece
    d[0x12] = Za_piece
    d[0x13] = Zb_piece

    d[0x14] = Sa_piece
    d[0x15] = Sb_piece
    d[0x16] = Sa_piece
    d[0x17] = Sb_piece

    d[0x18] = Ta_piece
    d[0x19] = Tb_piece
    d[0x1a] = Tc_piece
    d[0x1b] = Td_piece

    return d[piece]
end

function get_next_piece_rotation(piece :: Int64)
    d = Dict{Int64, Int64}()

    d[0x00] = 0x03
    d[0x01] = 0x00
    d[0x02] = 0x01
    d[0x03] = 0x02

    d[0x04] = 0x07
    d[0x05] = 0x04
    d[0x06] = 0x05
    d[0x07] = 0x06

    d[0x08] = 0x0b
    d[0x09] = 0x08
    d[0x0a] = 0x09
    d[0x0b] = 0x0a

    d[0x0c] = 0x0f
    d[0x0d] = 0x0c
    d[0x0e] = 0x0d
    d[0x0f] = 0x0e

    d[0x10] = 0x13
    d[0x11] = 0x10
    d[0x12] = 0x11
    d[0x13] = 0x12

    d[0x14] = 0x17
    d[0x15] = 0x14
    d[0x16] = 0x15
    d[0x17] = 0x16

    d[0x18] = 0x1b
    d[0x19] = 0x18
    d[0x1a] = 0x19
    d[0x1b] = 0x1a

    return d[piece]
end

function init_game_state()
    brain.mutation_rate  = 0.02
    brain.crossover_rate = 0.85

    brain.number_of_features = 4
    brain.fields_per_feature = 2
    brain.population_size    = 20

    brain.current_generation = 0
    brain.individual_index   = 1

    brain.number_of_evaluations_per_agent      = 5
    brain.number_of_evaluations_per_agent_left = brain.number_of_evaluations_per_agent

    brain.function_evaluations_total = 100
    brain.function_evaluations_left  = brain.function_evaluations_total

    range = 5.0

    brain.population = rand(brain.population_size,
                            brain.number_of_features * brain.fields_per_feature) .* (range * 2.0) .- range

    brain.agent_data = zeros(Float64, (brain.population_size, brain.number_of_evaluations_per_agent))

    brain.fitness = zeros(brain.population_size)
end

function frame_update()
    update_game_state()
    extract_board()
end

function check_for_new_piece()
    if garapa_read_pc() == 0x2062 && gs.screen == ingame
        new_piece()
    end

    #=if garapa_read_pc() == 0x2a0e=#
        #=input_handler(gs, brain)=#
    #=end=#
end

function new_piece()
    #=println("New Piece")=#
    gs.new_piece = true
    #=find_best_move(gs)=#
end

function garapa_init()
    println("Starting GABATE")
    garapa_register_callback_display_update(frame_update)
    garapa_register_callback_cpu_loop(check_for_new_piece)
end

function entered_game()
    #=println("Entered a GAME")=#
    gs.new_piece = true
end

function game_over()
    #=println("Game over")=#

    brain.number_of_evaluations_per_agent_left -= 1
    brain.function_evaluations_left            -= 1

    if brain.number_of_evaluations_per_agent_left <= 0
        brain.number_of_evaluations_per_agent_left = brain.number_of_evaluations_per_agent
        brain.individual_index += 1

        if brain.individual_index > brain.population_size
            evo_step()
            brain.individual_index = 1
        end
    end
end

function evo_step()
    evaluate_fitness()
end

function evaluate_fitness()
    for i in 1:brain.population_size
        brain.fitness[i] = mean(brain.agent_data[1:1:brain.number_of_evaluations_per_agent])
    end
end

function finished_boot()
    init_game_state()
    #=println("BOOT")=#
end

function entered_menu()
    #=println("MENU")=#
end

function draw_overlay()
    posx, posy = 400, 10

    text = @sprintf("screen: 0x%04x", garapa_read_byte(0xffe1))
    garapa_draw_test_with_bg(text, posx, posy, 0, 255, 255)
    posy += 20

    text = @sprintf("gen: %4d - %4d - %4d", brain.current_generation, brain.individual_index, brain.number_of_evaluations_per_agent_left)
    garapa_draw_test_with_bg(text, posx, posy, 0, 255, 255)
    posy += 20

    #=text = @sprintf("generation: %4d", brain.current_generation)=#
    #=garapa_draw_test_with_bg(text, posx, posy, 0, 255, 255)=#
    #=posy += 20=#

    if gs.screen == ingame
        px = (garapa_read_byte(0xff92) - 8 ) * 4
        py = (garapa_read_byte(0xff93) - 16) * 4

        piece = get_piece_offsets(garapa_read_byte(0xc203))

        garapa_draw_rectangle(px + piece[1, 1] * 4 * 8, py + piece[1, 2] * 4 * 8, 8 * 4, 8 * 4, 0  , 0  , 255)
        garapa_draw_rectangle(px + piece[2, 1] * 4 * 8, py + piece[2, 2] * 4 * 8, 8 * 4, 8 * 4, 0  , 0  , 255)
        garapa_draw_rectangle(px + piece[3, 1] * 4 * 8, py + piece[3, 2] * 4 * 8, 8 * 4, 8 * 4, 0  , 0  , 255)
        garapa_draw_rectangle(px + piece[4, 1] * 4 * 8, py + piece[4, 2] * 4 * 8, 8 * 4, 8 * 4, 127, 127, 255)

        best_piece = get_piece_offsets(gs.best_piece)

        garapa_draw_rectangle((gs.best_x + 1 + best_piece[1, 1]) * 4 * 8, (gs.best_y + best_piece[1, 2]) * 4 * 8, 8 * 4, 8 * 4, 0  , 255, 0  )
        garapa_draw_rectangle((gs.best_x + 1 + best_piece[2, 1]) * 4 * 8, (gs.best_y + best_piece[2, 2]) * 4 * 8, 8 * 4, 8 * 4, 0  , 255, 0  )
        garapa_draw_rectangle((gs.best_x + 1 + best_piece[3, 1]) * 4 * 8, (gs.best_y + best_piece[3, 2]) * 4 * 8, 8 * 4, 8 * 4, 0  , 255, 0  )
        garapa_draw_rectangle((gs.best_x + 1 + best_piece[4, 1]) * 4 * 8, (gs.best_y + best_piece[4, 2]) * 4 * 8, 8 * 4, 8 * 4, 127, 255, 127)
    end
end

function piece_moved()

end

olds = 0xff

function update_game_state()
    input_handler(gs, brain)

    if gs.new_piece
        gs.new_piece = false
        bx, by, rotation, best_piece = find_best_move(gs, brain)

        gs.best_x = bx
        gs.best_y = by
        gs.best_r = rotation
        gs.best_piece = best_piece
    end

    draw_overlay()
    screen = garapa_read_byte(0xffe1)

    #=if olds != screen=#
        #=@printf("%02x %02x\n", olds, screen)=#
    #=end=#

    global olds = screen

    px = div(garapa_read_byte(0xff92) - 8 , 8)
    py = div(garapa_read_byte(0xff93) - 16, 8)

    if px != gs.px || py != gs.py
        if screen == 0x00
            piece_moved()
        end

        gs.px = px
        gs.py = py
    end

    old_screen = gs.screen

    if screen == 0x00
        global gs.screen = ingame
    elseif screen == 0xff
        global gs.screen = boot
    elseif screen == 0x24 || screen == 0x25
        global gs.screen = credits
    elseif screen == 0x06 || screen == 0x07
        global gs.screen = menu
    elseif screen == 0x08 || screen == 0x0e
        global gs.screen = gametype
    elseif screen == 0x10 || screen == 0x11 || screen == 0x0a
        global gs.screen = atype
    elseif screen == 0x00
        global gs.screen = ingame
    elseif screen == 0x01 || screen == 0x0d || screen == 0x04
        global gs.screen = gameover
    end

    if old_screen != gs.screen
        if old_screen == ingame && gs.screen == gameover
            game_over()
        elseif old_screen != ingame && gs.screen == ingame
            entered_game()
        elseif old_screen == boot && gs.screen != boot
            finished_boot()
        elseif old_screen != menu && gs.screen == menu
            entered_menu()
        end
    end
end

function extract_board()
    x = 0
    y = 0
    for base in 0xc822:0x20:0xca22
        y += 1
        for offset in 0x00:0x09
            x += 1
            address = base + offset

            byte = garapa_read_byte(address)

            occupied = (byte == 0x2f ? false : true)
            global gs.board[y, x] = occupied
        end

        x = 0
    end
end

